vm/devices/firmware/firmware_uefi/src/service/diagnostics.rs

```diff
@@ -0,0 +1,476 @@
+    let base_offset = size_of::<AdvancedLoggerMessageEntryV2>()
+        .checked_add(message_len as usize)
+        .ok_or_else(|| {
+            EntryParseError::Overflow(anyhow!(
+                "size_of::<AdvancedLoggerMessageEntryV2> ({}) + message_len ({})",
+                size_of::<AdvancedLoggerMessageEntryV2>(),
+                message_len
+            ))
+        })?;
+
+    // Add padding for 8-byte alignment
+    let aligned_offset = base_offset.checked_add(ALIGNMENT_MASK).ok_or_else(|| {
+        EntryParseError::Overflow(anyhow!(
+            "base_offset ({}) + ALIGNMENT_MASK ({})",
+            base_offset,
+            ALIGNMENT_MASK
+        ))
+    })?;
+    let entry_size = aligned_offset & !ALIGNMENT_MASK;
+
+    Ok(EntryData {
+        debug_level: entry.debug_level,
+        time_stamp: entry.time_stamp,
+        phase: entry.phase,
+        message,
+        entry_size,
+    })
+}
+
+/// Errors that occur during processing
+#[derive(Debug, Error)]
+#[allow(missing_docs)]
+pub enum DiagnosticsError {
+    #[error("Failed to parse entry: {0}")]
+    EntryParse(#[from] EntryParseError),
+    #[error("Expected: {0:#x}, got: {1:#x}")]
+    HeaderSignatureMismatch(u32, u32),
+    #[error("Expected log buffer size < {0:#x}, got: {1:#x}")]
+    HeaderBufferSize(u32, u32),
+    #[error("Bad GPA value: {0:#x}")]
+    BadGpa(u32),
+    #[error("No GPA set")]
+    NoGpa,
+    #[error("Failed to read from guest memory: {0}")]
+    GuestMemoryRead(#[from] GuestMemoryError),
+    #[error("Encountered arithmetic overflow: {0}")]
+    Overflow(#[from] anyhow::Error),
+    #[error("Expected used log buffer size < {0:#x}, got: {1:#x}")]
+    BadUsedBufferSize(u32, u32),
+    #[error("Expected accumulated message length < {0:#x}, got: {1:#x}")]
+    BadAccumulatedMessageLength(u16, u16),
+}
+
+/// Definition of the diagnostics services state
+#[derive(Inspect)]
+pub struct DiagnosticsServices {
+    /// The guest physical address of the diagnostics buffer
+    gpa: Option<u32>,
+    /// Flag to indicate if we have already processed the buffer
+    did_process: bool,
+}
+
+impl DiagnosticsServices {
+    /// Create a new instance of the diagnostics services
+    pub fn new() -> DiagnosticsServices {
+        DiagnosticsServices {
+            gpa: None,
+            did_process: false,
+        }
+    }
+
+    /// Reset the diagnostics services state
+    pub fn reset(&mut self) {
+        self.gpa = None;
+        self.did_process = false;
+    }
+
+    /// Set the GPA of the diagnostics buffer
+    pub fn set_gpa(&mut self, gpa: u32) {
+        self.gpa = match gpa {
+            0 => None,
+            _ => Some(gpa),
+        }
+    }
+
+    /// Process the diagnostics buffer
+    pub fn process_diagnostics<F>(
+        &self,
+        gm: &GuestMemory,
+        mut log_handler: F,
+    ) -> Result<(), DiagnosticsError>
+    where
+        F: FnMut(EfiDiagnosticsLog),
+    {
+        // Step 1: Validate GPA
```

Comment:
Don't enumerate steps like this. That's just more work for the next person to edit this code when they add or remove a step--now they have to renumber everything (and they won't).